test/test_mqtt.py

```diff
@@ -459,8 +462,76 @@ def on_connection_closed_callback(connection, callback_data: OnConnectionClosedD
             exception_occurred = True
         self.assertTrue(exception_occurred, "Exception did not occur when connecting with invalid arguments!")
 
-        failureData = onConnectionFailureFuture.result(TIMEOUT)
-        self.assertTrue(failureData['error'] is not None)
+        failure_data = on_onnection_failure_future.result(TIMEOUT)
+        self.assertTrue(failure_data['error'] is not None)
+
+    def test_connect_disconnect_with_callbacks_happy_on_resume(self):
+        # Check that an on_connection_success callback fires on a resumed connection.
+
+        # NOTE Since there is no mocked server available on this abstraction level, the only sensible approach
+        # is to interrupt a connection, and wait for it to be resumed automatically. For that, another client
+        # with the same client_id connects to the server and then immediately disconnects.
+
+        test_input_endpoint = _get_env_variable("AWS_TEST_MQTT311_IOT_CORE_HOST")
+        test_input_cert = _get_env_variable("AWS_TEST_MQTT311_IOT_CORE_RSA_CERT")
+        test_input_key = _get_env_variable("AWS_TEST_MQTT311_IOT_CORE_RSA_KEY")
+        test_tls_opts = TlsContextOptions.create_client_with_mtls_from_path(test_input_cert, test_input_key)
+        test_tls = ClientTlsContext(test_tls_opts)
+
+        on_connection_success_future = Future()
+        on_connection_closed_future = Future()
+        on_connection_resumed_future = Future()
+
+        def on_connection_success_callback(connection, callback_data: OnConnectionSuccessData):
+            on_connection_success_future.set_result(
+                {'return_code': callback_data.return_code, "session_present": callback_data.session_present})
+
+        def on_connection_closed_callback(connection, callback_data: OnConnectionClosedData):
+            on_connection_closed_future.set_result({})
+
+        def on_connection_resumed_callback(connection, return_code: ConnectReturnCode, session_present):
+            on_connection_resumed_future.set_result(
+                {'return_code': return_code, "session_present": session_present})
+
+        connection = self._create_connection(
+            endpoint=test_input_endpoint,
+            tls_context=test_tls,
+            on_connection_success_callback=on_connection_success_callback,
+            on_connection_closed_callback=on_connection_closed_callback,
+            on_connection_resumed_callback=on_connection_resumed_callback)
+        connection.connect().result(TIMEOUT)
+        success_data = on_connection_success_future.result(TIMEOUT)
+        self.assertEqual(success_data['return_code'], ConnectReturnCode.ACCEPTED)
+        self.assertEqual(success_data['session_present'], False)
+
+        on_connection_success_future = Future()
```

Comment:
I'd add a commentary note RE: "reset the futures for the reconnect attempt"  Took me a few minutes to realize the reuse wasn't a mistake.